src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,125 @@
+                    + size);
+        }
+        grow();
+        if (size == index) {
+            internalArray[size] = value;
+        } else {
+            Object[] internalArrayTemp = new Object[internalArray.length];
+            System.arraycopy(internalArray,0, internalArrayTemp, 0, index);
+            System.arraycopy(internalArray,index, internalArrayTemp, index + 1, size - index);
+            internalArrayTemp[index] = value;
+            internalArray = internalArrayTemp;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            throw new NullPointerException("List<> shouldn't be null");
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) internalArray[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        internalArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        if (index < internalArray.length) {
+            Object[] internalArrayTemp = new Object[internalArray.length];
+            size = size - 1;
+            System.arraycopy(internalArray,0, internalArrayTemp, 0, index);
+            System.arraycopy(internalArray,index + 1, internalArrayTemp, index,
+                    internalArray.length - index - 1);
+            T removableElement = (T) internalArray[index];
+            internalArray = internalArrayTemp;
+            return removableElement;
+        }
+        throw new ArrayListIndexOutOfBoundsException("Element with index " + index
+                + " doesn't exist. Size of array is only " + size);
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        Object[] internalArrayTemp = new Object[internalArray.length];
+        for (int i = 0; i < size; i++) {
+            if (element == (T) internalArray[i] || (element != null
+                    && element.equals((T) internalArray[i]))) {
+                System.arraycopy(internalArray,0, internalArrayTemp, 0, i);
+                System.arraycopy(internalArray,i + 1, internalArrayTemp, i,
+                        size - i - 1);
+                T removeThisElement = (T) internalArray[i];
+                size = size - 1;
+                internalArray = internalArrayTemp;
+                return removeThisElement;
+            }
+        }
+        throw new NoSuchElementException("Element " + element + " doesn't exist.");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
```

Comment:
This method doesn't always grows. First it checks length. Better name for this method would be: "checkLengthAndGrow"